src/v/coproc/script_context_router.cc

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "script_context_router.h"
+
+#include "reflection/absl/node_hash_map.h"
+
+namespace reflection {
+
+ss::future<>
+async_adl<coproc::read_context>::to(iobuf& out, coproc::read_context&& rctx) {
+    reflection::serialize(
+      out, rctx.absolute_start, rctx.last_read, rctx.last_acked);
+    return ss::now();
+}
+
+ss::future<coproc::read_context>
+async_adl<coproc::read_context>::from(iobuf_parser& in) {
+    coproc::read_context rctx;
+    rctx.absolute_start = adl<model::offset>{}.from(in);
+    rctx.last_read = adl<model::offset>{}.from(in);
+    rctx.last_acked = adl<model::offset>{}.from(in);
+    return ss::make_ready_future<coproc::read_context>(std::move(rctx));
+}
+
+ss::future<>
+async_adl<coproc::write_context>::to(iobuf& out, coproc::write_context&& wctx) {
+    return async_adl<coproc::write_context::offsets_t>{}.to(
+      out, std::move(wctx.offsets));
+}
+
+ss::future<coproc::write_context>
+async_adl<coproc::write_context>::from(iobuf_parser& in) {
+    return async_adl<coproc::write_context::offsets_t>{}.from(in).then(
+      [](coproc::write_context::offsets_t result) {
+          return coproc::write_context{.offsets = std::move(result)};
+      });
+}
+
+ss::future<> async_adl<ss::lw_shared_ptr<coproc::source>>::to(
+  iobuf& out, ss::lw_shared_ptr<coproc::source>&& src) {
+    /// Take copies of the data the shared_ptr points to
+    auto rctx_cp = src->rctx;
+    auto wctx_cp = src->wctx;
+    adl<int8_t>{}.to(out, src->version);
```

Comment:
how does the addition of a version here relate to this comment suggesting its not needed? https://github.com/vectorizedio/redpanda/pull/2866#discussion_r752355278 is the data going on disk or over the network or both? what about the other structures that are serialized? if you're adding the version, do you need to use it in a verification procedure?

Reply:
I didn't say it wasn't needed just that now i have no use for it at the moment. I would add it now Incase in the future we need to expand the data types.

This router type is serialized to disk only to record the current offsets for copro it's never sent over the network to the wasm engine.

Good point about verification

Reply:
> I would add it now Incase in the future we need to expand the data types.

this is the reason for the version. but you'll have to make the call if it is needed or not and which types.